Add documentation for UI development (#65)
smartdevicelink/sdl_core_guides: docs/Developing the HMI UI/index.md

!!!

## Switching Templates

SDL Core will request the HMI to change an apps template from a `UI.Show` request, or the deprecated RPC `UI.SetDisplayLayout`.

The `UI.Show` request has a parameter `templateConfiguration` which includes a string for the requested layout.

The `UI.SetDisplayLayout` request includes a string parameter, `displayLayout`.

`UI.Show` is meant to be the preferred method because a `UI.Show` request can change the layout of the screen and the screen contents in a single request. This helps prevent lag and screen flashing when an app wants to change an app template. 

An SDL app should only request to view templates that are supported in the HMI Capabilities. The HMI may return a failed response to Core in the event an unsupported template is requested. 

More on HMI capabilities will be discussed later in this guide.

## Supported Template Views

A reference list for all supported template views can be found [here](https://smartdevicelink.com/en/guides/sdl-overview-guides/user-interface/supported-templates/).

This list shows screen shots of the 15 supported template views and how their text, graphic, and soft button components are layed out.

> The defined strings for each template can be found in `PredefinedLayout` struct in the Mobile API RPC Specification. 

Author: PredefinedLayout does not exist in the hmi api

## Creating the App Menu

The in App Menu should be made accessible from an app's template view. Please note the example placement of the hamburger menu icon in the top right of this screen shot.

![Menu Icon](./assets/menu_circled.jpg)

!!! NOTE
If the user chooses to open the menu, the HMI must send a `UI.OnSystemContext` notification with the `SystemContext` enum: `MENU`. After the user exits the menu, another `UI.OnSystemContext` notification must be sent with the `SystemContext` value: `MAIN`.
!!!

The contents of the HMI's menu are populated by the RPC `UI.AddCommand`. Each `UI.AddCommand` received corresponds to an individual menu item. When the user selects a menu item, the HMI should send a `UI.OnCommand` notification. We believe it is best practice to exit the menu after a user makes a selection from the list of commands.

There is some minor customization for the app menu. An HMI can choose to implement the app menu in a tile view, list view, or both. If an app has a preference for a type of menu layout, the HMI will receive a `UI.SetGlobalProperties` request from SDL Core.

> ![List Menu](./assets/list_menu.png)

Reviewer: Also, could help to include headers before the images, as of now they are just images appearing the page
```suggestion
#### List Menu Example

![List Menu](./assets/list_menu.png)
```

![Tiles Menu](./assets/tiles_menu.png)

## Implementing Popups

> There are several rpcs that are meant to display a popup or an overlay to the user.

Reviewer: ```suggestion
There are several RPCs which are used to display a popup or an overlay to the user.
```

> ### `UI.Alert`

Reviewer: These headers show up oddly in the dev portal since because of the backticks:
```suggestion
### UI.Alert
```

Same goes for the other headers in this section.

> Alert is a simple popup that can contain an image, text, and buttons.

Reviewer: ```suggestion
Alert is used to display a simple popup that can contain an image, text, and buttons.
```

>  ![Alert](./assets/alert.png)

Reviewer: This image is abnormally large

Author: Not sure what to do about that, seems like a dev portal style issue? All of the other pictures are screenshots i took but the alert and list view screenshots are the only images that adjust to the screen width.

### `UI.PerformInteraction`

> `PerformInteraction` is a ppoup with contents displayed similar to the app menu

Reviewer: ```suggestion
`PerformInteraction` is used to display a popup with contents which are displayed in a similar way to the app menu.
```

![Perform Interaction](./assets/perform_interaction.png)

### `UI.Slider`

> `Slider` is a popup that allows the user to enter an input via a slider button.

Reviewer: ```suggestion
`Slider` is used to display a popup that allows the user to enter a value via a slider input.
```

![Slider](./assets/slider.png)

### `UI.ScrollableMessage`

> ![Scrollable Message](./assets/scrollable_message.png)

Reviewer: Should the image be after the description, as is the case with the rest of the RPCs?

> `ScrollableMessage` is a popup that allows the user to read a longer message that requires scrolling.

Reviewer: ```suggestion
`ScrollableMessage` is used to display a popup which shows a long message to the user that requires scrolling.
```

!!! NOTE

> It is important that the HMI sends SDL Core a `UI.OnSystemContext` notification when displaying and then hiding a popup. `UI.Alert` is the only popup to have it's own SystemContext enum `ALERT`. All other popups must use the SystemContext enum value, `HMI_OBSCURED`.

Reviewer: ```suggestion
It is important that the HMI sends SDL Core a `UI.OnSystemContext` notification when displaying and closing a popup. A `systemContext` value of `ALERT` is used when `UI.Alert` is active, `HMI_OBSCURED` is used for all other popups.
```

!!!

## Navigating Through the IVI

> It is common for an SDL UI to be integrated into an existing OEM's UI. In order for SDL Core to work well with a head unit that has other embedded components, the HMI should make use of the `BasicCommunication.OnEventChanged` notification. This notification allows connected SDL applications to receive updates about their HMI statuses when a user interacts with other components like the embedded navigation or radio. For example if an SDL media application has been activated and is playing audio, and a user selects to switch the audio source to the embedded radio, the HMI should send SDL Core a `BasicCommunication.OnEventChanged` notification with `eventName = AUDIO_SOURCE` and `isActive = true`. This HMI notification will let the media application know that it no longer has control of the audio source and its SDL capabilities are overwritten by the embedded audio's capabilities.

Reviewer: ```suggestion
It is common for an SDL UI to be integrated into an existing OEM's UI. In order for SDL Core to work well with a head unit that has other embedded components, the HMI should make use of the `BasicCommunication.OnEventChanged` notification. This notification allows connected SDL applications to receive updates about their HMI status when a user interacts with other components like the embedded navigation or radio. For example, if an SDL media application is active and is playing audio, then the user switches the audio source to the embedded radio, the HMI should send SDL Core a `BasicCommunication.OnEventChanged` notification with `eventName = AUDIO_SOURCE` and `isActive = true`. This HMI notification will let the media application know that it no longer has control of the audio source.
```
Not certain what to replace `and its SDL capabilities are overwritten by the embedded audio's capabilities` with, but I don't feel the wording is quite clear there.

> If the user selects the media app as the audio source again, the HMI should send the same `BasicCommunication.OnEventChanged` notification, but with `isActive = false`. This will indicate to SDL Core that the application has control of the audio again.

Reviewer: ```suggestion
If the user selects the media app as the audio source again, the HMI should send the same `BasicCommunication.OnEventChanged` notification, but with `isActive = false`. This will indicate to SDL Core that the application has regained control of the audio.
```

## Defining the UI Capabilities

> There are several ways that the HMI should communicate its UI capabilities to SDL Core. When first connecting the HMI to SDL Core, SDL Core will send a `UI.GetCapabilities` request (a similar GetCapabilities request is sent for every interface). The HMI's response should include accurate information relating to supported display capabilities, audio pass through capabilities, soft button capabilities, and other system capabilities. 

Reviewer: ```suggestion
There are several ways that the HMI should communicate its UI capabilities to SDL Core. When first connecting the HMI to SDL Core, SDL Core will send a `UI.GetCapabilities` request (a similar GetCapabilities request is sent for every interface). The HMI's response should include accurate information relating to its supported display capabilities, audio pass through capabilities, soft button capabilities, and various other system capabilities (See [UI.GetCapabilities](https://smartdevicelink.com/en/guides/hmi/ui/getcapabilities/#parameters_1)).
```

> It is likely that the UI capabilities will be different for each template view, therefore it is important for the HMI to send updates about its capabilities to SDL Core. For example, if an app requests a new template configuration, after switching to that view, the HMI must send an `OnSystemCapabilityUpdated` notification for the `SystemCapability` type, `DISPLAYS`.  

Reviewer: ```suggestion
It is likely that the UI capabilities will be different for each template view, therefore it is important for the HMI to send updates about its capabilities to SDL Core. For example, if an app requests a new template configuration, after switching to that view the HMI must send an `OnSystemCapabilityUpdated` notification for `"systemCapabilityType": "DISPLAYS"`.  
```

!!! NOTE

> Even though `UI.SetDisplayLayout` is deprecated, the HMI can still receive requests to change layouts from this RPC. It is important that the response for `UI.SetDisplayLayout` includes accurate display capabilities for the new layout.

Reviewer: ```suggestion
Even though `UI.SetDisplayLayout` is deprecated, the HMI can still receive requests to change layouts from this RPC from older applications. It is important that the response for `UI.SetDisplayLayout` includes accurate display capabilities for the new layout as well.
```
